crates/polars-parquet/src/parquet/read/page/memreader.rs

```diff
@@ -149,33 +166,104 @@ impl io::Seek for MemReader {
             },
         };
 
-        eprintln!(
-            "pos = {}, new_pos = {}, seek = {:?}",
-            self.position, position, pos
-        );
-
         self.position = position;
 
         Ok(position as u64)
     }
 }
 
-impl MemReaderSlice {
+impl MemSlice {
     #[inline(always)]
     pub fn to_vec(self) -> Vec<u8> {
-        <[u8]>::to_vec(&self)
+        <[u8]>::to_vec(self.deref())
     }
 
     #[inline]
     pub fn from_vec(v: Vec<u8>) -> Self {
         let end = v.len();
 
-        Self {
-            data: v.into(),
+        Self(MemSliceInner::Allocated(AllocatedSlice {
+            data: v.into_boxed_slice().into(),
             start: 0,
             end,
+        }))
+    }
+
+    #[inline]
+    pub fn from_slice(slice: &[u8]) -> Self {
+        let end = slice.len();
+        Self(MemSliceInner::Allocated(AllocatedSlice {
+            data: slice.into(),
+            start: 0,
+            end,
+        }))
+    }
+
+    #[inline]
+    pub fn populate(&self) {
+        #[cfg(target_arch = "x86_64")]
+        {
+            if self.len() == 0 {
+                return;
+            }
+
+            const PREFETCH: unsafe fn(*const u8) = {
+                #[cfg(target_arch = "x86_64")]
```

Comment:
How does this work? Does it pre-fetch but doesn't block the thread?

Reply:
Yes, it sends an hint instruction to start fetching that memory already. The processor is allowed to ignore it, but usually it will at least try.